torchvision/prototype/transforms/functional/_misc.py

```diff
@@ -69,6 +69,10 @@ def normalize(
                 f"but got {type(inpt)} instead."
             )
 
+    # Saw this from [Feedback] thread (https://github.com/pytorch/vision/issues/6753#issuecomment-1308295943)
+    # Not sure I understand the need to return a Tensor instead of and Image
+    # Is the inconsistency "worth" it? How can we make sure this isn't too unexpected?
```

Comment:
Thanks for the details!

For me to better understand the trade-off, I'll try to find reasons to keep the output as Image instead of tensor. LMK what you think.

> we have the tensor to image fallback

I guess that also means one can do e.g. `Compose(Normalize(), SomeColorTransform())` and while Normalize will output a float tensor, `SomeColorTransform()` will treat it as a `0-1` image? Basically, the unwrapping may not be preventing any bug?

I'm also wondering whether this might become obsolete / overkill if we remove the `ColorSpace` metadata, which as discussed previously is currently redundant with `num_channels`?

> Of course any ops not related to color like crop will still work as before.

Do we have transforms that rely on the ColorSpace meta-data right now? I could only find `RandomPhotometricDistort` (and it can probably do without)

Reply:
> I guess that also means one can do e.g. `Compose(Normalize(), SomeColorTransform())` and while Normalize will output a float tensor, `SomeColorTransform()` will treat it as a `0-1` image? Basically, the unwrapping may not be preventing any bug?

Yes and yes. This is more of an idealistic choice. Value range checks are too expensive at runtime and this we can only assume the range. Meaning, this unwrapping (or rather the absent re-wrapping) will basically have no effect at runtime.

> I'm also wondering whether this might become obsolete / overkill if we remove the `ColorSpace` metadata, which as discussed previously is currently redundant with `num_channels`?

I don't think so. `datapoints.Image` (and `datapoints.Video`) carry more implicit assumptions than just the public metadata. An `Image` instance implicitly communicates that its values are in the correct range, whereas `torch.Tensor` is completely free. Of course if one uses it as image the same implicit assumptions apply, but this comes from the context rather than the type.

> Do we have transforms that rely on the ColorSpace meta-data right now? I could only find `RandomPhotometricDistort` (and it can probably do without)

For reference:

https://github.com/pytorch/vision/blob/93df9a50885d0345e31bba691576c83d5cee7737/torchvision/prototype/transforms/_color.py#L127-L128

First of all, it is an oversight that we are using two different idioms. Both `Normalize` and `RandomPhotometricDistort` should either return a `torch.Tensor` or an `Image` with `color_space=ColorSpace.OTHER`. 

To answer the question, yes it can do without. Apart from that, `ConvertColorSpace` also uses it, but we could also guess it from the channels. Although that would drop the guard that we currently have of transforming arbitrary "images", i.e. the normalized ones, to a different color space. Whether this is a good thing is probably open for discussion.

Reply:
> Both Normalize and RandomPhotometricDistort should either return a torch.Tensor or an Image with color_space=ColorSpace.OTHER.

I think the reason for `RandomPhotometricDistort` to return an Image is that adjusting hue, saturation, etc should not go outside the original color space (RGB -> RGB), however `Normalize` returns a tensor due to the fact that float Image is supposed to be between [0, 1] but output normalized tensor can have any kind of ranges...

Reply:
> should not go outside the original color space (RGB -> RGB), 


But then why setting it to `ColorSpace.OTHER` instead of keeping as `ColorSpace.RGB`?

It seems like there are 2 different concepts:

- The color-space of an image: RGB vs RGBA vs greyscale (... vs. YUV vs CMYK which aren't supported in torchvision)
- The range of values, and the current assumption that it is `[0, 1.0 if image.is_floating_point() else torch.iinfo(image.dtype).max]`

Conceptually these are 2 different things to me. E.g. I would argue that an RGB image in [0-1] is **still** an RGB image even after it is normalized: the range assumption may be broken, but it's still an RGB image - just in a non-standard range. Not sure I'm convincing but:

- `normalize` (out = (X - mean) / std) is just a generalization of `convertImageDtype` (out = X / 255), and `convertImageDtype` preserves the `ColorSpace` attribute.
- passing an RGB image to `normalize`  changes its range, but it doesn't convert it to a greyscale or RGBA or YUV / CMYK.

So I think there's an argument to be made to return Images instead of Tensors after `normalize`: I agree that the assumption about the range is broken, but a) this assumption isn't encoded in the ColorSpace anyway (it's encoded in the dtype) and b) returning a Tensor doesn't prevent any more issue than returning an Image.

Nor sure if I'm too clear but happy to chat more about this!

Reply:
> So I think there's an argument to be made to return Images instead of Tensors after normalize: I agree that the assumption about the range is broken, but a) this assumption isn't encoded in the ColorSpace anyway (it's encoded in the dtype) and b) returning a Tensor doesn't prevent any more issue than returning an Image.

Fair point. I can get behind that.

Reply:
I agree, these are colorspace and data range are two different things but there are relations between them especially when we have to convert from one colorspace to another. For example, RGB to HSV is well-defined for [0-256] and [0-1] ranges, produced HSV image has predefined ranges for 8-bit, 16-bit or float32 dtypes (ref [famous opencv link](https://docs.opencv.org/4.6.0/de/d25/imgproc_color_conversions.html)).
So, I think once RGB image is normalized with `mean, std` we can't convert it correctly to HSV space or to Gray scale...

Reply:
It's definitely true that these 2 concepts start to overlap when we're talking about more color-spaces like HSV / CYMK / YUV etc.

But these aren't supported in torchvision, and there is no immediate plan to support them either (I'm not saying never, but still). So at this point in time, the distinction may be somewhat premature and may lock us into a design that we might never actually need?

Reply:
That's true that we do not have HSV and other colorspaces right now and I agree that we probably wont support them since tomorrow. However, we are already doing RGB to grayscale and adjusting hue/saturation by internally doing rgb to hsv conversion. I'm happy to return Image from `Normalize` if it help with API adoption etc but we have to pay attention on output correctness and potential issues this step could bring...

Reply:
Yes this is definitely valid, we should also make sure that returing an Image doesn't lock us either. Do we foresee any hard issue by returning an Image?

Reply:
> Do we foresee any hard issue by returning an Image?

I don't. Like I stated above, this was an idealistic choice in the first place. Due to our fallback, it made no difference at runtime whatsoever.